In [1]:
# Reading in the initial data

import pandas as pd

matches = pd.read_csv("matches.csv", index_col=0)

In [2]:
matches.head()

date   time            comp        round  day venue result   gf   ga  \
1  2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2  2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3  2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4  2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6  2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  match report  notes    sh   sot  dist   fk   pk pkatt  \
1       Tottenham  ...  Match Report    NaN  18.0   4.0  16.9  1.0  0.0   0.0   
2    Norwich City  ...  Match Report    NaN  16.0   4.0  17.3  1.0  0.0   0.0   
3         Arsenal  ...  Match Report    NaN  25.0  10.0  14.3  0.0  0.0   0.0   
4  Leicester City  ...  Match Report    NaN  25.0   8.0  14.0  0.0  0.0   0.0   
6     Southampton  ...  Match Report    NaN  16.0   1.0  15.7  1.0  0.0   0.0   

   season             team  
1    2022  Manchester City  
2    2022  Manchester City  
3    2022  Manchester City  
4    2022  Manchester City  
6    2022  Manchester City  

[5 rows x 27 columns]

In [3]:
matches.shape

(1389, 27)

### Investigating Missing Data

We would like to investigate the number of matches that each team played over the season. We do know that every team plays 38 matches in the league each season. Hence `38 * 20 * 2 = 1520` is the total number of games expected in the data set. First let's check the number of matches that each team has played.

In [4]:
matches['team'].value_counts()

Southampton                 72
Brighton and Hove Albion    72
Manchester United           72
West Ham United             72
Newcastle United            72
Burnley                     71
Leeds United                71
Crystal Palace              71
Manchester City             71
Wolverhampton Wanderers     71
Tottenham Hotspur           71
Arsenal                     71
Leicester City              70
Chelsea                     70
Aston Villa                 70
Everton                     70
Liverpool                   38
Fulham                      38
West Bromwich Albion        38
Sheffield United            38
Brentford                   34
Watford                     33
Norwich City                33
Name: team, dtype: int64

Lets see how many matches were played each week

In [5]:
matches['round'].value_counts()

Matchweek 1     39
Matchweek 16    39
Matchweek 34    39
Matchweek 32    39
Matchweek 31    39
Matchweek 29    39
Matchweek 28    39
Matchweek 26    39
Matchweek 25    39
Matchweek 24    39
Matchweek 23    39
Matchweek 2     39
Matchweek 19    39
Matchweek 17    39
Matchweek 20    39
Matchweek 15    39
Matchweek 5     39
Matchweek 3     39
Matchweek 13    39
Matchweek 12    39
Matchweek 4     39
Matchweek 11    39
Matchweek 10    39
Matchweek 9     39
Matchweek 8     39
Matchweek 14    39
Matchweek 7     39
Matchweek 6     39
Matchweek 30    37
Matchweek 27    37
Matchweek 22    37
Matchweek 21    37
Matchweek 18    37
Matchweek 33    32
Matchweek 35    20
Matchweek 36    20
Matchweek 37    20
Matchweek 38    20
Name: round, dtype: int64

The reason we have some weeks having fewer games is because we scrapped the date during the 21/22 season, which means some of the games were still ongoing, that why we have some weeks having fewer than 39 matches. This also explains why the data has fewer number of rows than what was expected for a full premier league season.

### Cleaning Data

We will start by checking the data types that we have. This is because machine learning algorithms can only work with numbers. They can't work with objects.

In [6]:
matches.dtypes

date             object
time             object
comp             object
round            object
day              object
venue            object
result           object
gf              float64
ga              float64
opponent         object
xg              float64
xga             float64
poss            float64
attendance      float64
captain          object
formation        object
referee          object
match report     object
notes           float64
sh              float64
sot             float64
dist            float64
fk              float64
pk              float64
pkatt           float64
season            int64
team             object
dtype: object

In [7]:
# convert date column into datetime using pandas
matches['date'] = pd.to_datetime(matches['date'])

In [8]:
matches.dtypes

date            datetime64[ns]
time                    object
comp                    object
round                   object
day                     object
venue                   object
result                  object
gf                     float64
ga                     float64
opponent                object
xg                     float64
xga                    float64
poss                   float64
attendance             float64
captain                 object
formation               object
referee                 object
match report            object
notes                  float64
sh                     float64
sot                    float64
dist                   float64
fk                     float64
pk                     float64
pkatt                  float64
season                   int64
team                    object
dtype: object

### Creating Predictors

We will create a new predictor by converting the Home/Away venue into a numerical data type

In [9]:
matches["venue_code"] = matches["venue"].astype("category").cat.codes

In [11]:
matches["opp_code"] = matches["opponent"].astype("category").cat.codes

In [13]:
matches["hour"] = matches["time"].str.replace(":.+", "", regex=True).astype("int")

In [14]:
matches["day_code"] = matches["date"].dt.dayofweek

In [16]:
matches["target"] = (matches['result'] == 'W').astype("int")

In [17]:
matches

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...   fk   pk  pkatt  season              team  \
1   1.0       Tottenham  ...  1.0  0.0    0.0    2022   Manchester City   
2   0.0    Norwich City  ...  1.0  0.0    0.0    2022   Manchester City   
3   0.0         Arsenal  ...  0.0  0.0    0.0    2022   Manchester City   
4   0.0  Leicester City  ...  0.0  0.0    0.0    2022   Manchester City   
6   0.0     Southampton  ...  1.0  0.0    0.0    2022   Manchester City   
..  ...             ...  ...  ...  ...    ...     ...               ...   
38  4.0       Tottenham  ...  0.0  0.0    0.0    2021  Sheffield United   
39  2.0  Crystal Palace  ...  1.0  0.0    0.0    2021  Sheffield United   
40  0.0         Everton  ...  0.0  0.0    0.0    2021  Sheffield United   
41  1.0   Newcastle Utd  ...  1.0  0.0    0.0    2021  Sheffield United   
42  0.0         Burnley  ...  0.0  0.0    0.0    2021  Sheffield United   

   venue_code opp_code hour  day_code  target  
1           0       18   16         6       0  
2           1       15   15         5       1  
3           1        0   12         5       1  
4           0       10   15         5       1  
6           1       17   15         5       0  
..        ...      ...  ...       ...     ...  
38          0       18   19         6       0  
39          1        6   15         5       0  
40          0        7   19         6       1  
41          0       14   18         2       0  
42          1        4   16         6       1  

[1389 rows x 32 columns]

### Creating Machine Learning Model

We will be using a classification algorithm called the Random Forest Classifier. This algorithms is able to pick up non-linearities in data sets better than linear models.

In [18]:
from sklearn.ensemble import RandomForestClassifier

In [19]:
rf = RandomForestClassifier(n_estimators=50, min_samples_split=10, random_state=1)

In [20]:
train = matches[matches['date'] < '2022-01-01']

In [21]:
test = matches[matches['date'] > '2022-01-01']

In [22]:
predictors = ["venue_code", "opp_code", "hour", "day_code"]

In [23]:
rf.fit(train[predictors], train["target"])

RandomForestClassifier(min_samples_split=10, n_estimators=50, random_state=1)

In [24]:
preds = rf.predict(test[predictors])

We would like to determine the accuracy of our model, with a confusion matrix

In [25]:
from sklearn.metrics import accuracy_score

In [26]:
acc = accuracy_score(test['target'], preds)

In [27]:
acc

0.6123188405797102

In [28]:
combined = pd.DataFrame(dict(actual=test['target'], prediction=preds))

In [29]:
pd.crosstab(index=combined["actual"], columns=combined["prediction"])

prediction    0   1
actual             
0           141  31
1            76  28

Let's test the precision of the model

In [30]:
from sklearn.metrics import precision_score

In [31]:
precision_score(test["target"], preds)

0.4745762711864407

### Improving precision with rolling averages

In [32]:
grouped_matches = matches.groupby("team")

In [33]:
group = grouped_matches.get_group("Manchester City")

In [35]:
group.head(3) # A group of manchester city matches 

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   

       opponent  ...   fk   pk  pkatt  season             team venue_code  \
1     Tottenham  ...  1.0  0.0    0.0    2022  Manchester City          0   
2  Norwich City  ...  1.0  0.0    0.0    2022  Manchester City          1   
3       Arsenal  ...  0.0  0.0    0.0    2022  Manchester City          1   

  opp_code hour  day_code  target  
1       18   16         6       0  
2       15   15         5       1  
3        0   12         5       1  

[3 rows x 32 columns]

In [37]:
def rolling_averages(group, cols, new_cols):
    group = group.sort_values("date")
    rolling_stats = group[cols].rolling(3, closed='left').mean()
    group[new_cols] = rolling_stats
    group = group.dropna(subset=new_cols)
    return group

In [38]:
cols = ["gf", "ga", "sh", "sot", "dist", "fk", "pk", "pkatt"]
new_cols = [f"{c}_rolling" for c in cols]
new_cols

['gf_rolling',
 'ga_rolling',
 'sh_rolling',
 'sot_rolling',
 'dist_rolling',
 'fk_rolling',
 'pk_rolling',
 'pkatt_rolling']

In [39]:
rolling_averages(group, cols, new_cols)

date   time            comp         round  day venue result   gf  \
5  2020-10-17  17:30  Premier League   Matchweek 5  Sat  Home      W  1.0   
7  2020-10-24  12:30  Premier League   Matchweek 6  Sat  Away      D  1.0   
9  2020-10-31  12:30  Premier League   Matchweek 7  Sat  Away      W  1.0   
11 2020-11-08  16:30  Premier League   Matchweek 8  Sun  Home      D  1.0   
12 2020-11-21  17:30  Premier League   Matchweek 9  Sat  Away      L  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
42 2022-03-14  20:00  Premier League  Matchweek 29  Mon  Away      D  0.0   
44 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Away      W  2.0   
46 2022-04-10  16:30  Premier League  Matchweek 32  Sun  Home      D  2.0   
49 2022-04-20  20:00  Premier League  Matchweek 30  Wed  Home      W  3.0   
50 2022-04-23  15:00  Premier League  Matchweek 34  Sat  Home      W  5.0   

     ga        opponent  ...  day_code  target  gf_rolling  ga_rolling  \
5   0.0         Arsenal  ...         5       1    2.000000    2.333333   
7   1.0        West Ham  ...         5       0    1.333333    2.000000   
9   0.0   Sheffield Utd  ...         5       1    1.000000    0.666667   
11  1.0       Liverpool  ...         6       0    1.000000    0.333333   
12  2.0       Tottenham  ...         5       0    1.000000    0.666667   
..  ...             ...  ...       ...     ...         ...         ...   
42  0.0  Crystal Palace  ...         0       0    2.333333    1.333333   
44  0.0         Burnley  ...         5       1    1.666667    0.333333   
46  2.0       Liverpool  ...         6       0    2.000000    0.333333   
49  0.0        Brighton  ...         2       1    1.333333    0.666667   
50  1.0         Watford  ...         5       1    2.333333    0.666667   

   sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
5   17.333333    4.666667    18.900000   1.333333    0.333333       0.333333  
7   17.333333    3.666667    17.733333   0.666667    0.000000       0.000000  
9   16.666667    4.333333    18.233333   0.666667    0.000000       0.000000  
11  14.333333    6.666667    18.466667   1.000000    0.000000       0.000000  
12  12.000000    5.666667    19.366667   1.000000    0.000000       0.333333  
..        ...         ...          ...        ...         ...            ...  
42  19.000000    7.000000    15.366667   0.333333    0.333333       0.333333  
44  18.333333    7.333333    16.000000   0.333333    0.000000       0.000000  
46  20.000000    6.666667    16.133333   0.333333    0.000000       0.000000  
49  15.666667    4.666667    16.700000   0.333333    0.000000       0.000000  
50  15.333333    5.000000    17.200000   0.333333    0.000000       0.000000  

[68 rows x 40 columns]

In [40]:
matches_rolling = matches.groupby('team').apply(lambda x: rolling_averages(x, cols, new_cols))

In [41]:
matches_rolling

date   time            comp         round  \
team                                                                         
Arsenal                 6  2020-10-04  14:00  Premier League   Matchweek 4   
                        7  2020-10-17  17:30  Premier League   Matchweek 5   
                        9  2020-10-25  19:15  Premier League   Matchweek 6   
                        11 2020-11-01  16:30  Premier League   Matchweek 7   
                        13 2020-11-08  19:15  Premier League   Matchweek 8   
...                               ...    ...             ...           ...   
Wolverhampton Wanderers 32 2022-03-13  14:00  Premier League  Matchweek 29   
                        33 2022-03-18  20:00  Premier League  Matchweek 30   
                        34 2022-04-02  15:00  Premier League  Matchweek 31   
                        35 2022-04-08  20:00  Premier League  Matchweek 32   
                        36 2022-04-24  14:00  Premier League  Matchweek 34   

                            day venue result   gf   ga         opponent  ...  \
team                                                                     ...   
Arsenal                 6   Sun  Home      W  2.0  1.0    Sheffield Utd  ...   
                        7   Sat  Away      L  0.0  1.0  Manchester City  ...   
                        9   Sun  Home      L  0.0  1.0   Leicester City  ...   
                        11  Sun  Away      W  1.0  0.0   Manchester Utd  ...   
                        13  Sun  Home      L  0.0  3.0      Aston Villa  ...   
...                         ...   ...    ...  ...  ...              ...  ...   
Wolverhampton Wanderers 32  Sun  Away      W  1.0  0.0          Everton  ...   
                        33  Fri  Home      L  2.0  3.0     Leeds United  ...   
                        34  Sat  Home      W  2.0  1.0      Aston Villa  ...   
                        35  Fri  Away      L  0.0  1.0    Newcastle Utd  ...   
                        36  Sun  Away      L  0.0  1.0          Burnley  ...   

                            day_code  target  gf_rolling  ga_rolling  \
team                                                                   
Arsenal                 6          6       1    2.000000    1.333333   
                        7          5       0    1.666667    1.666667   
                        9          6       0    1.000000    1.666667   
                        11         6       1    0.666667    1.000000   
                        13         6       0    0.333333    0.666667   
...                              ...     ...         ...         ...   
Wolverhampton Wanderers 32         6       1    1.333333    1.000000   
                        33         4       0    1.666667    0.666667   
                        34         5       1    2.333333    1.000000   
                        35         4       0    1.666667    1.333333   
                        36         6       0    1.333333    1.666667   

                           sh_rolling sot_rolling dist_rolling fk_rolling  \
team                                                                        
Arsenal                 6    7.666667    3.666667    14.733333   0.666667   
                        7    5.333333    3.666667    15.766667   0.000000   
                        9    7.000000    3.666667    16.733333   0.666667   
                        11   9.666667    4.000000    16.033333   1.000000   
                        13   9.666667    2.666667    18.033333   1.000000   
...                               ...         ...          ...        ...   
Wolverhampton Wanderers 32  12.333333    3.666667    19.300000   0.000000   
                        33  12.333333    4.333333    19.600000   0.000000   
                        34  13.000000    5.333333    19.833333   0.000000   
                        35  13.000000    5.000000    18.533333   0.000000   
                        36  10.000000    4.666667    17.633333   0.000000   

                            pk_rolling

In [42]:
matches_rolling = matches_rolling.droplevel('team')

In [43]:
matches_rolling

date   time            comp         round  day venue result   gf  \
6  2020-10-04  14:00  Premier League   Matchweek 4  Sun  Home      W  2.0   
7  2020-10-17  17:30  Premier League   Matchweek 5  Sat  Away      L  0.0   
9  2020-10-25  19:15  Premier League   Matchweek 6  Sun  Home      L  0.0   
11 2020-11-01  16:30  Premier League   Matchweek 7  Sun  Away      W  1.0   
13 2020-11-08  19:15  Premier League   Matchweek 8  Sun  Home      L  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
32 2022-03-13  14:00  Premier League  Matchweek 29  Sun  Away      W  1.0   
33 2022-03-18  20:00  Premier League  Matchweek 30  Fri  Home      L  2.0   
34 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
35 2022-04-08  20:00  Premier League  Matchweek 32  Fri  Away      L  0.0   
36 2022-04-24  14:00  Premier League  Matchweek 34  Sun  Away      L  0.0   

     ga         opponent  ...  day_code  target  gf_rolling  ga_rolling  \
6   1.0    Sheffield Utd  ...         6       1    2.000000    1.333333   
7   1.0  Manchester City  ...         5       0    1.666667    1.666667   
9   1.0   Leicester City  ...         6       0    1.000000    1.666667   
11  0.0   Manchester Utd  ...         6       1    0.666667    1.000000   
13  3.0      Aston Villa  ...         6       0    0.333333    0.666667   
..  ...              ...  ...       ...     ...         ...         ...   
32  0.0          Everton  ...         6       1    1.333333    1.000000   
33  3.0     Leeds United  ...         4       0    1.666667    0.666667   
34  1.0      Aston Villa  ...         5       1    2.333333    1.000000   
35  1.0    Newcastle Utd  ...         4       0    1.666667    1.333333   
36  1.0          Burnley  ...         6       0    1.333333    1.666667   

   sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
6    7.666667    3.666667    14.733333   0.666667    0.000000       0.000000  
7    5.333333    3.666667    15.766667   0.000000    0.000000       0.000000  
9    7.000000    3.666667    16.733333   0.666667    0.000000       0.000000  
11   9.666667    4.000000    16.033333   1.000000    0.000000       0.000000  
13   9.666667    2.666667    18.033333   1.000000    0.333333       0.333333  
..        ...         ...          ...        ...         ...            ...  
32  12.333333    3.666667    19.300000   0.000000    0.000000       0.000000  
33  12.333333    4.333333    19.600000   0.000000    0.000000       0.000000  
34  13.000000    5.333333    19.833333   0.000000    0.000000       0.000000  
35  13.000000    5.000000    18.533333   0.000000    0.000000       0.000000  
36  10.000000    4.666667    17.633333   0.000000    0.000000       0.000000  

[1317 rows x 40 columns]

In [44]:
matches_rolling.index = range(matches_rolling.shape[0])

In [45]:
matches_rolling

date   time            comp         round  day venue result   gf  \
0    2020-10-04  14:00  Premier League   Matchweek 4  Sun  Home      W  2.0   
1    2020-10-17  17:30  Premier League   Matchweek 5  Sat  Away      L  0.0   
2    2020-10-25  19:15  Premier League   Matchweek 6  Sun  Home      L  0.0   
3    2020-11-01  16:30  Premier League   Matchweek 7  Sun  Away      W  1.0   
4    2020-11-08  19:15  Premier League   Matchweek 8  Sun  Home      L  0.0   
...         ...    ...             ...           ...  ...   ...    ...  ...   
1312 2022-03-13  14:00  Premier League  Matchweek 29  Sun  Away      W  1.0   
1313 2022-03-18  20:00  Premier League  Matchweek 30  Fri  Home      L  2.0   
1314 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
1315 2022-04-08  20:00  Premier League  Matchweek 32  Fri  Away      L  0.0   
1316 2022-04-24  14:00  Premier League  Matchweek 34  Sun  Away      L  0.0   

       ga         opponent  ...  day_code  target  gf_rolling  ga_rolling  \
0     1.0    Sheffield Utd  ...         6       1    2.000000    1.333333   
1     1.0  Manchester City  ...         5       0    1.666667    1.666667   
2     1.0   Leicester City  ...         6       0    1.000000    1.666667   
3     0.0   Manchester Utd  ...         6       1    0.666667    1.000000   
4     3.0      Aston Villa  ...         6       0    0.333333    0.666667   
...   ...              ...  ...       ...     ...         ...         ...   
1312  0.0          Everton  ...         6       1    1.333333    1.000000   
1313  3.0     Leeds United  ...         4       0    1.666667    0.666667   
1314  1.0      Aston Villa  ...         5       1    2.333333    1.000000   
1315  1.0    Newcastle Utd  ...         4       0    1.666667    1.333333   
1316  1.0          Burnley  ...         6       0    1.333333    1.666667   

     sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
0      7.666667    3.666667    14.733333   0.666667    0.000000       0.000000  
1      5.333333    3.666667    15.766667   0.000000    0.000000       0.000000  
2      7.000000    3.666667    16.733333   0.666667    0.000000       0.000000  
3      9.666667    4.000000    16.033333   1.000000    0.000000       0.000000  
4      9.666667    2.666667    18.033333   1.000000    0.333333       0.333333  
...         ...         ...          ...        ...         ...            ...  
1312  12.333333    3.666667    19.300000   0.000000    0.000000       0.000000  
1313  12.333333    4.333333    19.600000   0.000000    0.000000       0.000000  
1314  13.000000    5.333333    19.833333   0.000000    0.000000       0.000000  
1315  13.000000    5.000000    18.533333   0.000000    0.000000       0.000000  
1316  10.000000    4.666667    17.633333   0.000000    0.000000       0.000000  

[1317 rows x 40 columns]

#### Retrain our model

Now that we have some new predictors with different stats and feature engineering. We can retrain our model to see if it improves

In [48]:
def make_predicitions(data, predictors):
    train = data[data['date'] < '2022-01-01']
    test = data[data['date'] > '2022-01-01']
    rf.fit(train[predictors], train["target"])
    preds = rf.predict(test[predictors])
    combined = pd.DataFrame(dict(actual=test['target'], prediction=preds))
    precision = precision_score(test["target"], preds)
    return combined, precision

In [49]:
combined, precision = make_predicitions(matches_rolling, predictors + new_cols)

In [50]:
precision

0.625

We see that our precision has improved to 62%

In [51]:
combined

actual  prediction
55         0           0
56         1           0
57         1           0
58         1           1
59         1           1
...      ...         ...
1312       1           0
1313       0           0
1314       1           0
1315       0           0
1316       0           0

[276 rows x 2 columns]

In [52]:
combined = combined.merge(matches_rolling[['date', 'team', 'opponent', 'result']], left_index=True, right_index=True)

In [53]:
combined

actual  prediction       date                     team       opponent  \
55         0           0 2022-01-23                  Arsenal        Burnley   
56         1           0 2022-02-10                  Arsenal         Wolves   
57         1           0 2022-02-19                  Arsenal      Brentford   
58         1           1 2022-02-24                  Arsenal         Wolves   
59         1           1 2022-03-06                  Arsenal        Watford   
...      ...         ...        ...                      ...            ...   
1312       1           0 2022-03-13  Wolverhampton Wanderers        Everton   
1313       0           0 2022-03-18  Wolverhampton Wanderers   Leeds United   
1314       1           0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
1315       0           0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
1316       0           0 2022-04-24  Wolverhampton Wanderers        Burnley   

     result  
55        D  
56        W  
57        W  
58        W  
59        W  
...     ...  
1312      W  
1313      L  
1314      W  
1315      L  
1316      L  

[276 rows x 6 columns]

In [54]:
class MissingDict(dict):
    __missing__ = lambda self, key: key

map_values = {
    "Brighton and Hove Albion": "Brighton",
    "Manchester United": "Manchester Utd",
    "Newcastle United": "Newcastle Utd",
    "Tottenham Hotspur": "Tottenham",
    "West Ham United": "West Ham",
    "Wolverhampton Wanderers": "Wolves"
}

mapping = MissingDict(**map_values)

In [55]:
mapping["West Ham United"]

'West Ham'

In [57]:
combined["new_team"] = combined['team'].map(mapping)

In [58]:
combined

actual  prediction       date                     team       opponent  \
55         0           0 2022-01-23                  Arsenal        Burnley   
56         1           0 2022-02-10                  Arsenal         Wolves   
57         1           0 2022-02-19                  Arsenal      Brentford   
58         1           1 2022-02-24                  Arsenal         Wolves   
59         1           1 2022-03-06                  Arsenal        Watford   
...      ...         ...        ...                      ...            ...   
1312       1           0 2022-03-13  Wolverhampton Wanderers        Everton   
1313       0           0 2022-03-18  Wolverhampton Wanderers   Leeds United   
1314       1           0 2022-04-02  Wolverhampton Wanderers    Aston Villa   
1315       0           0 2022-04-08  Wolverhampton Wanderers  Newcastle Utd   
1316       0           0 2022-04-24  Wolverhampton Wanderers        Burnley   

     result new_team  
55        D  Arsenal  
56        W  Arsenal  
57        W  Arsenal  
58        W  Arsenal  
59        W  Arsenal  
...     ...      ...  
1312      W   Wolves  
1313      L   Wolves  
1314      W   Wolves  
1315      L   Wolves  
1316      L   Wolves  

[276 rows x 7 columns]

In [59]:
merged = combined.merge(combined, left_on=["date", "new_team"], right_on=["date", "opponent"])

In [60]:
merged

actual_x  prediction_x       date                   team_x  \
0           0             0 2022-01-23                  Arsenal   
1           1             0 2022-02-10                  Arsenal   
2           1             0 2022-02-19                  Arsenal   
3           1             1 2022-02-24                  Arsenal   
4           1             1 2022-03-06                  Arsenal   
..        ...           ...        ...                      ...   
257         1             0 2022-03-13  Wolverhampton Wanderers   
258         0             0 2022-03-18  Wolverhampton Wanderers   
259         1             0 2022-04-02  Wolverhampton Wanderers   
260         0             0 2022-04-08  Wolverhampton Wanderers   
261         0             0 2022-04-24  Wolverhampton Wanderers   

        opponent_x result_x new_team_x  actual_y  prediction_y  \
0          Burnley        D    Arsenal         0             0   
1           Wolves        W    Arsenal         0             0   
2        Brentford        W    Arsenal         0             0   
3           Wolves        W    Arsenal         0             0   
4          Watford        W    Arsenal         0             0   
..             ...      ...        ...       ...           ...   
257        Everton        W     Wolves         0             0   
258   Leeds United        L     Wolves         1             0   
259    Aston Villa        W     Wolves         0             0   
260  Newcastle Utd        L     Wolves         1             0   
261        Burnley        L     Wolves         1             0   

                      team_y opponent_y result_y     new_team_y  
0                    Burnley    Arsenal        D        Burnley  
1    Wolverhampton Wanderers    Arsenal        L         Wolves  
2                  Brentford    Arsenal        L      Brentford  
3    Wolverhampton Wanderers    Arsenal        L         Wolves  
4                    Watford    Arsenal        L        Watford  
..                       ...        ...      ...            ...  
257                  Everton     Wolves        L        Everton  
258             Leeds United     Wolves        W   Leeds United  
259              Aston Villa     Wolves        L    Aston Villa  
260         Newcastle United     Wolves        W  Newcastle Utd  
261                  Burnley     Wolves        W        Burnley  

[262 rows x 13 columns]

In [61]:
merged[(merged['prediction_x'] == 1) & (merged['prediction_y'] == 0)]['actual_x'].value_counts()

1    27
0    13
Name: actual_x, dtype: int64

In [62]:
27/40

0.675

After feature engineering the model with additional stats, we successfully improved the model's predictive accuracy to 67.5%.